File: 10-retry-test-actions/java/src/test/java/RetryTestActions.java
package src.test.java;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;

/**
 * Created by andrew on 8/22/15.
 */
public class RetryTestActions {
    public class WorkWithBasicAuth {
        WebDriver driver;

        @Before
        public void setUp() throws Exception {
            driver = new FirefoxDriver();
        }

        @Test
        public void retryTestActionsTest() {
            driver.get("http://the-internet.herokuapp.com/notification_message");
            assert retryIfNotificationMessageContains("please try again");
        }

        @After
        public void tearDown() throws Exception {
            driver.quit();
        }


        public String getNotificationMessage() {
            return driver.findElement(By.id("flash")).getText();
        }

        public boolean retryIfNotificationMessageContains(String message) {
            for (int count = 0; count < 3; count++) {
                if (getNotificationMessage().contains(message))
                    driver.navigate().refresh();
                else
                    return true;
            }
            return false;
        }
    }
}

File: 10-retry-test-actions/kotlin/src/test/kotlin/RetryTestActions.kt
import org.openqa.selenium.By
import org.openqa.selenium.firefox.FirefoxDriver
import kotlin.test.Test
import kotlin.test.AfterTest
import kotlin.test.assertTrue

//10-retry-test-actions
class RetryTestActions {
    private val testUrl = "http://the-internet.herokuapp.com/notification_message"

    private val driver by lazy {
        FirefoxDriver()
    }

    @AfterTest
    fun tearDown() {
        // Quit session (closes browser)
        driver.quit()
    }

    @Test
    fun retryTestActionsTest() {
        driver.get(testUrl)
        assertTrue(retryIfNotificationMessageContains("please try again"))
    }

    private fun getNotificationMessage(): String = driver.findElement(By.id("flash")).text

    private fun retryIfNotificationMessageContains(message: String): Boolean {
        for (count in 0..2) {
            if (getNotificationMessage() == message)
                driver.navigate().refresh()
            else
                return true
        }
        return false
    }
}